File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/ParentObject.cpp
#include <ace.h>
#include <gtest/gtest.h>
#include <memory>
#include "../EngineTest.h"

using namespace std;
using namespace ace;

class ObjectSystem_ParentObject : public EngineTest
{
public:
	ObjectSystem_ParentObject(bool isOpenGLMode)
		: EngineTest(ace::ToAString("ParentObject"), isOpenGLMode, 50)
		, m_parent(nullptr)
	{
	}

protected:
	shared_ptr<TextureObject2D> m_parent;

	void OnStart()
	{
		auto scene = make_shared<Scene>();
		auto layer = make_shared<Layer2D>();
		auto child = make_shared<TextureObject2D>();
		auto child2 = make_shared<TextureObject2D>();
		m_parent = make_shared<TextureObject2D>();

		auto file = ace::Engine::GetFile();
		file->AddRootDirectories(ace::ToAString("Data/Texture").c_str());

		ace::Engine::ChangeScene(scene);
		scene->AddLayer(layer);
		layer->AddObject(m_parent);
		layer->AddObject(child);
		layer->AddObject(child2);
		m_parent->AddChild(child, ChildMode::All);
		child->AddChild(child2, ChildMode::Nothing);

		auto g = ace::Engine::GetGraphics();
		auto texture = g->CreateTexture2D(ace::ToAString("Cloud1.png").c_str());
		m_parent->SetTexture(texture);
		child->SetTexture(texture);
		child2->SetTexture(texture);

		m_parent->SetPosition(Vector2DF(320, 240));
		m_parent->SetCenterPosition(Vector2DF(128, 128));
		m_parent->SetColor(Color(255, 255, 128, 255));
		child->SetPosition(Vector2DF(50, 50));
		child2->SetPosition(Vector2DF(60, 60));
	}

	void OnUpdating()
	{
		m_parent->SetAngle(m_parent->GetAngle() + 0.2f);
		m_parent->SetScale(m_parent->GetScale() + Vector2DF(0.001f, 0.001f));
	}
};

ENGINE_TEST(ObjectSystem, ParentObject)

File: Dev/unitTest_Engine_cs/ObjectSystem2D/ParentObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.ObjectSystem2D
{
	class ParentObject : EngineTest
	{
		public ParentObject()
			: base( 60 )
		{
		}

		protected override void OnStart()
		{
			var scene = new Scene();
			var layer = new Layer2D();
			var child = new TextureObject2D();
			parent = new TextureObject2D();

			parent.AddChild( child, ChildMode.All );
			layer.AddObject( parent );
			layer.AddObject( child );
			scene.AddLayer( layer );
			Engine.ChangeScene( scene );

			var texture = Engine.Graphics.CreateTexture2D( CloudTexturePath );

			parent.Position = new Vector2DF( 320, 240 );
			parent.CenterPosition = new Vector2DF( 128, 128 );
			parent.Texture = texture;

			child.Position = new Vector2DF( 50, 50 );
			child.Texture = texture;
		}

		protected override void OnUpdating()
		{
			parent.Angle += 0.2f;
			parent.Scale = new Vector2DF() { X = parent.Scale.X + 0.0008f, Y = parent.Scale.Y + 0.0008f };
		}

		private TextureObject2D parent { get; set; }
	}
}
